File: src/route/step-round.ts
import type { Point } from '../layout/types'

const dist = (a: Point, b: Point) => Math.hypot(b.x - a.x, b.y - a.y)

function toward(from: Point, to: Point, d: number): Point {
  const len = dist(from, to) || 1
  return { x: from.x + ((to.x - from.x) / len) * d, y: from.y + ((to.y - from.y) / len) * d }
}

function collinear(a: Point, b: Point, c: Point): boolean {
  return Math.abs((b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x)) < 0.01
}

/**
 * Build an SVG path through a polyline, rounding each interior corner with a
 * quarter-circle (quadratic) arc. Radius is clamped to half of each adjacent
 * segment so tight turns stay clean. Generalizes the hand-tuned `step-round`
 * path from the source toolkit to an arbitrary turn-point list.
 */
export function roundedPath(points: Point[], radius: number): string {
  const pts = dedupe(points)
  if (pts.length < 2) return ''
  if (pts.length === 2) return `M ${r(pts[0].x)} ${r(pts[0].y)} L ${r(pts[1].x)} ${r(pts[1].y)}`

  let d = `M ${r(pts[0].x)} ${r(pts[0].y)}`
  for (let i = 1; i < pts.length - 1; i++) {
    const prev = pts[i - 1]
    const cur = pts[i]
    const next = pts[i + 1]
    if (collinear(prev, cur, next)) {
      d += ` L ${r(cur.x)} ${r(cur.y)}`
      continue
    }
    const rad = Math.min(radius, dist(prev, cur) / 2, dist(cur, next) / 2)
    const a = toward(cur, prev, rad)
    const b = toward(cur, next, rad)
    d += ` L ${r(a.x)} ${r(a.y)} Q ${r(cur.x)} ${r(cur.y)} ${r(b.x)} ${r(b.y)}`
  }
  const last = pts[pts.length - 1]
  d += ` L ${r(last.x)} ${r(last.y)}`
  return d
}

function dedupe(points: Point[]): Point[] {
  const out: Point[] = []
  for (const p of points) {
    const prev = out[out.length - 1]
    if (!prev || Math.abs(prev.x - p.x) > 0.5 || Math.abs(prev.y - p.y) > 0.5) out.push(p)
  }
  return out
}

const r = (n: number) => Math.round(n * 100) / 100

export { dist }

File: src/Beck/Route/StepRound.cs
using Beck.Layout;
using Beck.Model;

namespace Beck.Route;

/// <summary>
/// Quadratic corner-rounding path builder — a port of <c>src/route/step-round.ts</c>.
/// Each interior corner is rounded with a quarter-circle (quadratic) arc, radius
/// clamped to half of each adjacent segment. Coordinates round to 2 decimals with
/// JS <c>Math.round</c> semantics.
/// </summary>
internal static class StepRound
{
    public static double Dist(Point a, Point b)
    {
        double dx = b.X - a.X, dy = b.Y - a.Y;
        return Math.Sqrt(dx * dx + dy * dy);
    }

    private static Point Toward(Point from, Point to, double d)
    {
        double len = Dist(from, to);
        if (len == 0) len = 1;
        return new Point(from.X + (to.X - from.X) / len * d, from.Y + (to.Y - from.Y) / len * d);
    }

    private static bool Collinear(Point a, Point b, Point c) =>
        Math.Abs((b.X - a.X) * (c.Y - a.Y) - (b.Y - a.Y) * (c.X - a.X)) < 0.01;

    /// <summary>2-decimal coordinate formatting (JS <c>r(n) = Math.round(n*100)/100</c>).</summary>
    public static string R(double n) => Js.Str(Js.Round(n * 100) / 100);

    public static string RoundedPath(IReadOnlyList<Point> points, double radius)
    {
        var pts = Dedupe(points);
        if (pts.Count < 2) return "";
        if (pts.Count == 2) return $"M {R(pts[0].X)} {R(pts[0].Y)} L {R(pts[1].X)} {R(pts[1].Y)}";

        var d = $"M {R(pts[0].X)} {R(pts[0].Y)}";
        for (int i = 1; i < pts.Count - 1; i++)
        {
            Point prev = pts[i - 1], cur = pts[i], next = pts[i + 1];
            if (Collinear(prev, cur, next)) { d += $" L {R(cur.X)} {R(cur.Y)}"; continue; }
            double rad = Math.Min(radius, Math.Min(Dist(prev, cur) / 2, Dist(cur, next) / 2));
            Point a = Toward(cur, prev, rad);
            Point b = Toward(cur, next, rad);
            d += $" L {R(a.X)} {R(a.Y)} Q {R(cur.X)} {R(cur.Y)} {R(b.X)} {R(b.Y)}";
        }
        Point last = pts[^1];
        d += $" L {R(last.X)} {R(last.Y)}";
        return d;
    }

    private static List<Point> Dedupe(IReadOnlyList<Point> points)
    {
        var outp = new List<Point>();
        foreach (var p in points)
        {
            if (outp.Count == 0) { outp.Add(p); continue; }
            Point prev = outp[^1];
            if (Math.Abs(prev.X - p.X) > 0.5 || Math.Abs(prev.Y - p.Y) > 0.5) outp.Add(p);
        }
        return outp;
    }
}
